File: src/main/capture-geometry.ts
// Pure capture geometry — extracted from CaptureController so the redaction-
// adjacent rect math (a pixel of drift misplaces a marker or a crop) is unit-
// testable without electron / node-screenshots. The NATIVE crop stays in
// CaptureController; here we only compute rectangles.
import type { Point, Rect } from '../shared/project';

/** A focused-window descriptor (subset of get-windows' result we classify on). */
export type ActiveLike = { owner: { name: string }; title: string } | null | undefined;

// Windows shell host processes whose windows are huge/transparent and capture as
// a black swath — observed from get-windows on Windows 11 (sometimes the friendly
// name, sometimes the exe).
const SHELL_HOST_RE =
  /experience host|searchhost|shellexperiencehost|startmenuexperiencehost|searchapp|textinputhost|cortana/i;

/** Smallest rectangle containing both inputs. */
export function unionRect(a: Rect, b: Rect): Rect {
  const x = Math.min(a.x, b.x);
  const y = Math.min(a.y, b.y);
  const right = Math.max(a.x + a.width, b.x + b.width);
  const bottom = Math.max(a.y + a.height, b.y + b.height);
  return { x, y, width: right - x, height: bottom - y };
}

/** A box of half-size `620·scaleFactor` centered on a point — the generous,
 *  roughly symmetric crop used to frame a click together with any menu/dropdown
 *  it opened (menus flip up/left near screen edges, so the box is symmetric). */
export function clickBox(point: Point, scaleFactor: number): Rect {
  const half = Math.round(620 * (scaleFactor || 1));
  return { x: point.x - half, y: point.y - half, width: half * 2, height: half * 2 };
}

/**
 * Classify how to frame a click from the focused window:
 *  - 'window'     → a normal app window
 *  - 'region'     → an OS shell surface (taskbar, Start/Search, tray) → tight crop
 *  - 'fullscreen' → the desktop, or unknown/unidentified focus → whole monitor
 */
export function captureModeFor(active: ActiveLike): 'window' | 'region' | 'fullscreen' {
  if (!active) return 'fullscreen'; // unknown focus → full context, never a guessed crop
  const app = active.owner.name;
  const title = active.title;
  if (app === 'Windows Explorer' && title === 'Program Manager') return 'fullscreen'; // desktop
  if (app === 'Windows Explorer' && title.trim() === '') return 'region'; // taskbar / system tray
  if (SHELL_HOST_RE.test(app)) return 'region'; // Start / Search / Shell hosts
  return 'window';
}

/**
 * The crop rectangle (in the monitor's LOCAL pixel space) for a global-px region,
 * clamped to the monitor bounds. `mon` is the monitor's global-px position/size.
 * Pure geometry of CaptureController.cropToRegion (which does the native crop).
 */
export function cropRect(
  mon: { x: number; y: number; width: number; height: number },
  region: Rect,
): Rect {
  const lx = Math.round(region.x - mon.x);
  const ly = Math.round(region.y - mon.y);
  const cropX = Math.max(0, Math.min(lx, mon.width - 1));
  const cropY = Math.max(0, Math.min(ly, mon.height - 1));
  const cropW = Math.max(1, Math.min(lx + Math.round(region.width), mon.width) - cropX);
  const cropH = Math.max(1, Math.min(ly + Math.round(region.height), mon.height) - cropY);
  return { x: cropX, y: cropY, width: cropW, height: cropH };
}

File: dotnet/src/ShotAI.Core/Capture/CaptureGeometry.cs
using ShotAI.Core.Json;
using ShotAI.Core.Model;
using ShotAI.Core.Shell;

namespace ShotAI.Core.Capture;

/// <summary>
/// The capture's rectangle math (spec 02 2.8.2, <c>src/main/capture-geometry.ts</c> and the
/// inline crops of <c>CaptureController.ts</c>). Inputs are global physical pixels; a crop is in
/// the pixels of the monitor's image, whose global origin is the monitor's top-left. Every
/// rounding is <see cref="JsMath.Round"/> (INV-CAP-18): a pixel of drift misplaces a crop or a
/// click marker.
/// </summary>
public static class CaptureGeometry
{
    /// <summary>The smallest rectangle that holds both.</summary>
    public static Rect UnionRect(Rect a, Rect b)
    {
        var x = Math.Min(a.X, b.X);
        var y = Math.Min(a.Y, b.Y);
        var right = Math.Max(a.X + a.Width, b.X + b.Width);
        var bottom = Math.Max(a.Y + a.Height, b.Y + b.Height);
        return new Rect(x, y, right - x, bottom - y);
    }

    /// <summary>
    /// The box a menu selection is framed with: <see cref="CaptureConstants.ClickBoxHalf"/> times
    /// the scale either side of the point, symmetric because menus flip up and left near screen
    /// edges. A scale of 0 or NaN counts as 1, as JavaScript's <c>scaleFactor || 1</c> does.
    /// </summary>
    public static Rect ClickBox(Point point, double scaleFactor)
    {
        var half = JsMath.Round(CaptureConstants.ClickBoxHalf * OrOne(scaleFactor));
        return new Rect(point.X - half, point.Y - half, half * 2, half * 2);
    }

    /// <summary>
    /// <c>cropRect</c>, the menu and window paths' crop: the region clamped to the monitor. The
    /// right and bottom edges come from the unclamped left and top, so a region hanging off the
    /// left or top shrinks; the crop is never empty.
    /// </summary>
    public static PixelRect CropRect(Rect monitor, Rect region)
    {
        var lx = JsMath.Round(region.X - monitor.X);
        var ly = JsMath.Round(region.Y - monitor.Y);
        var cropX = Math.Max(0, Math.Min(lx, monitor.Width - 1));
        var cropY = Math.Max(0, Math.Min(ly, monitor.Height - 1));
        var cropW = Math.Max(1, Math.Min(lx + JsMath.Round(region.Width), monitor.Width) - cropX);
        var cropH = Math.Max(1, Math.Min(ly + JsMath.Round(region.Height), monitor.Height) - cropY);
        return Pixels(cropX, cropY, cropW, cropH);
    }

    /// <summary>
    /// The area path's crop, deliberately not <see cref="CropRect"/>: an area hanging off the
    /// monitor's left or top keeps its full size, extending from the edge (REQUIRED; the macOS
    /// port keeps the same quirk).
    /// </summary>
    public static PixelRect AreaCrop(Rect monitor, Rect area)
    {
        var cropX = Math.Max(0, Math.Min(JsMath.Round(area.X - monitor.X), monitor.Width - 1));
        var cropY = Math.Max(0, Math.Min(JsMath.Round(area.Y - monitor.Y), monitor.Height - 1));
        var cropW = Math.Max(1, Math.Min(JsMath.Round(area.Width), monitor.Width - cropX));
        var cropH = Math.Max(1, Math.Min(JsMath.Round(area.Height), monitor.Height - cropY));
        return Pixels(cropX, cropY, cropW, cropH);
    }

    /// <summary>
    /// The auto shell-region crop around a click: an 820 x 640 logical box centred on the point,
    /// no larger than the monitor, and shifted rather than shrunk to stay on it.
    /// </summary>
    public static PixelRect RegionCrop(Rect monitor, double scaleFactor, Point point)
    {
        var sf = OrOne(scaleFactor);
        var boxW = Math.Min(JsMath.Round(CaptureConstants.RegionBoxWidth * sf), monitor.Width);
        var boxH = Math.Min(JsMath.Round(CaptureConstants.RegionBoxHeight * sf), monitor.Height);
        var cx = point.X - monitor.X;
        var cy = point.Y - monitor.Y;
        var cropX = Math.Max(0, Math.Min(cx - Math.Floor(boxW / 2), monitor.Width - boxW));
        var cropY = Math.Max(0, Math.Min(cy - Math.Floor(boxH / 2), monitor.Height - boxH));
        return Pixels(cropX, cropY, boxW, boxH);
    }

    // JavaScript's `f || 1`: 0 and NaN are falsy.
    private static double OrOne(double f) => f == 0 || double.IsNaN(f) ? 1 : f;

    // The operands are whole pixels by the time they arrive here.
    private static PixelRect Pixels(double x, double y, double width, double height) =>
        new((int)x, (int)y, (int)width, (int)height);
}
